rapid/src/main/java/com/vrg/rapid/MembershipService.java

```diff
@@ -426,11 +484,48 @@ private void decideViewChange(final List<Endpoint> proposal) {
     /**
      * Invoked by observers of a node for failure detection.
      */
+    @SuppressWarnings("FutureReturnValueIgnored")
     private ListenableFuture<RapidResponse> handleMessage(final ProbeMessage probeMessage) {
         LOG.trace("handleProbeMessage from {}", Utils.loggable(probeMessage.getSender()));
+        final long observerConfigurationId = probeMessage.getObserverConfigurationId();
+
+        if (!membershipHistory.stream().anyMatch(change -> change.getConfigurationId() == observerConfigurationId)) {
```

Comment:
Given that this a list that we keep appending to, it would be faster to read it in reverse (a node that is missing updates is more likely to have missed recent ones -- so no need to scan all the way from the beginning of the history).

Reply:
This should be addressed now by using the `ConcurrentLinkedDeque` and prepending entries